File: src/doc_update_tool/diffing/diff_report.py
"""Render a ChangeRecord list as the "新舊對照列表" (old-vs-new comparison table) xlsx —
the same report format regardless of whether the source was Excel or Word."""

from __future__ import annotations

from pathlib import Path

import openpyxl

from doc_update_tool.models import ChangeRecord, ChangeType

_HEADERS = ["位置", "欄位名稱", "舊內容", "新內容", "變動類型"]


def write_diff_report(records: list[ChangeRecord], path: Path) -> None:
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "差異報告"
    ws.append(_HEADERS)

    for record in records:
        if record.change_type == ChangeType.UNCHANGED:
            continue  # internal bookkeeping only, not shown to the user
        ws.append(
            [
                record.location_id,
                record.field_name or "",
                record.old_zh or "",
                record.new_zh or "",
                record.change_type.value,
            ]
        )

    path.parent.mkdir(parents=True, exist_ok=True)
    wb.save(path)

File: web/src/diff-report.ts
/** Mirrors diffing/diff_report.py — renders a ChangeRecord list as the "新舊對照列表"
 * xlsx, built fresh each time so there's no round-trip fidelity risk here. */

import type ExcelJSNS from "exceljs";

import { ChangeType } from "./models.ts";
import type { ChangeRecord } from "./models.ts";
import { toArrayBuffer } from "./xlsx-buffer.ts";

const HEADERS = ["位置", "欄位名稱", "舊內容", "新內容", "變動類型"];

export async function buildDiffReportWorkbook(
  ExcelJS: typeof ExcelJSNS,
  records: ChangeRecord[]
): Promise<ArrayBuffer> {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet("差異報告");
  ws.addRow(HEADERS);

  for (const record of records) {
    if (record.changeType === ChangeType.UNCHANGED) continue; // internal bookkeeping only
    ws.addRow([record.locationId, record.fieldName ?? "", record.oldZh ?? "", record.newZh ?? "", record.changeType]);
  }

  return toArrayBuffer(await wb.xlsx.writeBuffer());
}

/** One combined report row across a whole batch of documents — unlike ChangeRecord,
 * this also carries which document it came from plus the user's on-screen note and
 * translation, since a batch run produces one merged report rather than one per file. */
export interface ReportRow {
  docName: string;
  locationId: string;
  fieldName: string | null;
  oldZh?: string;
  newZh?: string;
  changeType: ChangeType;
  note: string;
  translation: string;
}

const COMBINED_HEADERS = ["文件名稱", "位置", "欄位名稱", "舊內容", "新內容", "變動類型", "備註", "翻譯"];

export async function buildCombinedReportWorkbook(ExcelJS: typeof ExcelJSNS, rows: ReportRow[]): Promise<ArrayBuffer> {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet("差異報告");
  ws.addRow(COMBINED_HEADERS);

  for (const row of rows) {
    if (row.changeType === ChangeType.UNCHANGED) continue;
    ws.addRow([row.docName, row.locationId, row.fieldName ?? "", row.oldZh ?? "", row.newZh ?? "", row.changeType, row.note, row.translation]);
  }

  return toArrayBuffer(await wb.xlsx.writeBuffer());
}
